### PR TITLE
Adaptive surface loss weight annealing (cosine decay, 2 arms)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Surface-loss weighting is currently **fixed** at `--surface-loss-weight 2.0` throughout all training epochs. We hypothesize that a **cosine-annealed surface loss weight** — high early (strong geometry learning signal), decaying to low late (freeing gradient budget for volume generalization) — will reduce the large OOD vol_p gap (val≈3.57% vs test≈11.35%, a 3.2× gap that dominates ensemble error).

Rationale:
- Early in training, surface geometry (pressure + wall-shear topology) is the dominant learning signal. A higher surface weight here anchors the model to physics-plausible solutions.
- Later, the model needs to generalize its volume predictions. Reducing surface weight lets volume gradients dominate the final optimization phase, improving vol_p OOD generalization.
- This mirrors curriculum learning principles and is analogous to cosine annealing in LR schedules — neither a new idea in isolation, but untested on the surface/volume trade-off axis.

**This requires a code change** in `target/train.py`: add `--surface-loss-weight-final` flag and implement per-epoch cosine interpolation.

### Code change (target/train.py)

1. Add a new CLI flag (near the existing `--surface-loss-weight`):
```python
parser.add_argument("--surface-loss-weight-final", type=float, default=None,
    help="If set, cosine-anneal surface loss weight from --surface-loss-weight (start) to this value (end) over training.")
```

2. In the training loop, at the **start of each epoch** (before loss computation), update the surface loss weight. Find where `cfg.surface_loss_weight` (or equivalent) is used in the loss calculation and insert:
```python
if cfg.surface_loss_weight_final is not None:
    import math
    frac = epoch / max(cfg.epochs - 1, 1)  # 0.0 → 1.0 over epochs
    w_start = cfg.surface_loss_weight
    w_end = cfg.surface_loss_weight_final
    effective_surface_weight = w_end + (w_start - w_end) * 0.5 * (1 + math.cos(math.pi * frac))
    # override the loss weight for this epoch
    surface_loss_weight = effective_surface_weight
else:
    surface_loss_weight = cfg.surface_loss_weight
```

Make sure to log `surface_loss_weight` to W&B each epoch (e.g. `wandb.log({"train/surface_loss_weight": surface_loss_weight})`).

### Run: 4-epoch screen — two arms

Run both arms in parallel (4 GPUs each, or sequentially if needed):

**Arm A — aggressive decay (3.0 → 0.5):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --surface-loss-weight 3.0 --surface-loss-weight-final 0.5 \
  --volume-loss-weight 1.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32778" \
  --no-compile-model --use-surf-to-vol-xattn \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group adaptive-surface-loss-weighting \
  --wandb-name askeladd/surf-anneal-3.0-to-0.5
```

**Arm B — moderate decay (2.5 → 1.0):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.5 --surface-loss-weight-final 1.0 \
  --volume-loss-weight 1.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32778" \
  --no-compile-model --use-surf-to-vol-xattn \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group adaptive-surface-loss-weighting \
  --wandb-name askeladd/surf-anneal-2.5-to-1.0
```

### Kill gate (EP4)

If val_abupt (val_primary/abupt_axis_mean_rel_l2_pct) > **6.44%** at epoch 4 for both arms → close and report NEGATIVE. If either arm beats 6.44%, run the best arm to 13 epochs (full training) with `--lr-cosine-t-max 13 --epochs 13` and `--vol-points-schedule "0:16384:3:32778:6:49152:9:65536"`.

### What to log in W&B
Per epoch, please log:
- `train/surface_loss_weight` — confirm the schedule is being applied correctly
- `val_primary/abupt_axis_mean_rel_l2_pct` — primary gate metric
- `val_surface/surface_pressure_rel_l2_pct`
- `val_volume/volume_pressure_rel_l2_pct` — key OOD diagnostic
- `val_surface/wall_shear_rel_l2_pct`

### What to report
Report per-arm results at EP4. For each arm:
- val_abupt, val_vol_p, val_surface_p, val_wall_shear
- The W&B `surface_loss_weight` curve (confirm cosine shape)
- Whether vol_p shows improvement vs baseline (val≈3.57%, test≈11.35%)

Terminal result marker (when complete):
```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id-arm-a>","<run-id-arm-b>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<best_value>},"test_metric":{"name":"test_primary/abupt_axis_mean_rel_l2_pct","value":<test_value>}}
```

## Baseline

Current single-model SOTA (PR #823, run `ghh0s4ne`):
- **val_abupt = 6.4407%** ← gate to beat
- val_surface_pressure = ~3.86%
- val_volume_pressure = ~3.57% (but test_vol_p ≈ 11.35% — 3.2× OOD gap)
- val_wall_shear = ~6.87%

Ensemble SOTA (PR #880, K=6): val_abupt = **6.0289%**, test_abupt = 7.3693%

W&B baseline run: `ghh0s4ne` — [https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ghh0s4ne](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/ghh0s4ne)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: This PR is ready to implement and launch. Please implement the `--surface-loss-weight-final` flag and run both arms as specified in the PR body.

## Implementation

Add the following to `target/train.py` (in the config/args section):
```python
parser.add_argument('--surface-loss-weight-final', type=float, default=None,
                    help='If set, cosine-anneal surface_loss_weight from --surface-loss-weight to this value over training')
```

Then in the training loop (per-epoch, before computing the loss), interpolate:
```python
if cfg.surface_loss_weight_final is not None:
    import math
    frac = epoch / max(cfg.epochs - 1, 1)  # 0.0 → 1.0 over epochs
    w_start = cfg.surface_loss_weight
    w_end = cfg.surface_loss_weight_final
    effective_surface_weight = w_end + (w_start - w_end) * 0.5 * (1 + math.cos(math.pi * frac))
    surface_loss_weight = effective_surface_weight
else:
    surface_loss_weight = cfg.surface_loss_weight
```

Log the effective weight each epoch:
```python
wandb.log({'train/surface_loss_weight': surface_loss_weight})
```

## Run both arms in parallel (4-epoch screen, DDP8)

**Arm A — aggressive anneal (3.0 → 0.5):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --surface-loss-weight 3.0 --surface-loss-weight-final 0.5 \
  --volume-loss-weight 1.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32778" \
  --no-compile-model --use-surf-to-vol-xattn \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group adaptive-surface-loss-weighting \
  --wandb-name askeladd/surf-anneal-3.0-to-0.5
```

**Arm B — moderate anneal (2.5 → 1.0):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.5 --surface-loss-weight-final 1.0 \
  --volume-loss-weight 1.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32778" \
  --no-compile-model --use-surf-to-vol-xattn \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group adaptive-surface-loss-weighting \
  --wandb-name askeladd/surf-anneal-2.5-to-1.0
```

## EP4 Gate

val_abupt (`val_primary/abupt_axis_mean_rel_l2_pct`) ≤ **6.44%** at epoch 4.

- Both arms fail gate → close NEGATIVE
- Either arm passes → extend best arm to 13 epochs: `--lr-cosine-t-max 13 --epochs 13 --vol-points-schedule "0:16384:3:32778:6:49152:9:65536"`

## Key diagnostics to report at EP4

Per arm: val_abupt, val_vol_p, val_surface_p, val_wall_shear, and the `train/surface_loss_weight` curve (confirm cosine shape). Even if abupt doesn't improve vs baseline (6.4407%), a val_vol_p drop below ~3.57% is informative.

**Comment by morganmcg1** [comment]
ADVISOR: Flagging an anomaly in the current W&B data for run Arm A (6s8cq4tx). At EP2, the ABUPT is showing **27.8%** — far above the expected baseline of ~7-8% at EP2. This strongly suggests the run started from random initialization rather than from a pretrained checkpoint, or there is a bug in the implementation of the `--surface-loss-weight-final` flag.

Please verify:
1. Did the run start from the correct pretrained checkpoint? (W&B config shows `checkpoint: NONE` — was this intentional?)
2. Is the `--surface-loss-weight-final` flag correctly parsed and applied in the cosine annealing schedule?
3. Is the surface loss weight schedule integrated correctly into the training loop without disrupting other loss terms?

At 27.8% ABUPT at EP2, this run will almost certainly fail the EP4 kill gate of ≤6.44%. Please investigate and report back. If the implementation has a bug, fix it and restart with the correct checkpoint before EP4 so we don't waste the full 4-epoch compute budget.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #947 because CLOSE NEGATIVE — Invalid experiment setup: runs started from random initialization (checkpoint=NONE) rather than the pretrained checkpoint. The EP4 kill gate of ≤6.44% ABUPT is unreachable from random init (baseline best run achieved 6.44% using a checkpoint). At EP2, Arm A shows 27.8% ABUPT vs expected ~7-8% from a checkpoint start.

Additionally, the val logging appears to have stalled: only 1 val checkpoint has appeared (at EP1.98) despite the run reaching EP3.45, which prevents any meaningful gate evaluation.

**Root cause of the PR setup error:** The run commands in the PR body do not include a `--checkpoint` flag. This was an omission in the original PR instructions — the EP4 gate cannot be fairly evaluated without starting from the same baseline checkpoint as the reference run.

**Hypothesis verdict:** The adaptive surface loss weighting idea itself remains untested. The implementation appears to have been added (W&B config shows surface_loss_weight_final=0.5) but the random-init baseline makes it impossible to attribute any result to the new feature vs the training dynamics.

**Future work:** If the team wants to test adaptive surface loss weighting, the next PR should:
1. Start from the best available pretrained checkpoint
2. Use --validation-every 1 (already in instructions)
3. Confirm val metrics are logging correctly after EP2 before committing to full EP4 run.

**Comment by morganmcg1** [comment]
STUDENT askeladd:

Investigated the 27.83% concern. **The run is healthy — this is the EP1 (end of first epoch) validation, not EP2.** Off-by-one in epoch indexing.

## Evidence

- W&B run `6s8cq4tx` has **only 1 validation** logged so far, at `_step=10864`
- `total_estimated_steps=43456` for 4 epochs ⇒ 10864 steps per epoch ⇒ this is end-of-EP0 (0-indexed) = "EP1" in the trajectory tables
- The SOTA `ghh0s4ne` (PR #823 surface→volume xattn, 13ep) had **EP1 val_abupt=28.6344% at the same step=10864**. Our 27.83% is actually slightly better.
- Currently mid-EP1 training (`Epoch 2/4: 77% | 8409/10864`); next validation at step≈21728 will be the EP2 datapoint. SOTA had 8.15% there.

## Cosine schedule verified working

- step=0: `train/surface_loss_weight=3.0` (epoch 0 start, w_start) ✓
- step≈10864: transitioned to `train/surface_loss_weight=2.375` (epoch 1, frac=1/3, expected: 0.5 + 2.5·0.5·1.5 = 2.375) ✓
- Will transition to 1.125 at epoch 2, 0.5 at epoch 3 ✓

## Re: `checkpoint: NONE` in W&B config

This is **expected** — `train.py` has no `--checkpoint` / `--resume` flag for warm-starting from a previous run. The SOTA `ghh0s4ne` (PR #823) and all prior single-model SOTAs in BASELINE.md were trained from scratch under the same protocol. The W&B `checkpoint` field refers only to internal best-checkpoint saving during a single run.

## Plan

- Continue Arm A to EP4. ETA ~3-4 more hours (5685s/epoch × 2.3 remaining).
- Launch Arm B (2.5→1.0 anneal) sequentially after Arm A finishes — both PR commands use `--nproc_per_node=8`, can't run in parallel on this 8-GPU node without halving GPUs.
- Will report both arms' EP4 metrics with the cosine-shape confirmation.

Apologies for the confusion — the trajectory looks fine.

**Comment by morganmcg1** [comment]
STUDENT askeladd: Killed Arm A training (PID 671759) and the queued Arm B launcher (PID 694805) since this PR is closed and the GPUs are needed for new assignment PR #956. Arm A had reached EP3.45 of 4 epochs at the time of kill. No further action on this PR.
